### ISSUE TITLE
Var Recipes error on Unbounded or Generic Type Parameter

### ISSUE THREAD

**Issue by MBoegers** [openrewrite/rewrite-migrate-java#257]
## What version of OpenRewrite are you using?
I am using Moderne.io, current.

## How are you running OpenRewrite?
I am using Moderne.io and triggered [UseVar](https://docs.openrewrite.org/recipes/java/migrate/lang/usevar) Recipe.
And discovered the problem in various Spring Projects like [spring-shell](https://app.moderne.io/results/euAyh/details/eyJfX3R5cGVuYW1lIjoiR2l0SHViUmVwb3NpdG9yeSIsIm9yaWdpbiI6ImdpdGh1Yi5jb20iLCJwYXRoIjoic3ByaW5nLXByb2plY3RzL3NwcmluZy1oYXRlb2FzIiwiYnJhbmNoIjoibWFpbiJ9?onlyShowErrors=true&referrer=%2Fresults%2FeuAyh)

## What is the smallest, simplest way to reproduce the problem?

```java
class A {
    static <T> void generic() {
        List<T> lst = new ArrayList<>();
    }
    void unbound() {
        List<?> lst = new ArrayList<>();
    }
    void inception() {
        List<List<Object>> lst = new ArrayList<>();
    }
}
```

## What did you expect to see?
```java
class A {
  static <T> void generic() {
      var lst = new ArrayList<T>();
  }
  void unbound() {
      var lst = new ArrayList<>();
  }
  void inception() {
      var lst = new ArrayList<List<Object>>();
  }
}
```

### Generic
```
class org.openrewrite.java.tree.JavaType$GenericTypeVariable cannot be cast to class org.openrewrite.java.tree.JavaType$Class (org.openrewrite.java.tree.JavaType$GenericTypeVariable and org.openrewrite.java.tree.JavaType$Class are in unnamed module of loader 'app')
```

### Unbound
```
class org.openrewrite.java.tree.JavaType$GenericTypeVariable cannot be cast to class org.openrewrite.java.tree.JavaType$Class (org.openrewrite.java.tree.JavaType$GenericTypeVariable and org.openrewrite.java.tree.JavaType$Class are in unnamed module of loader 'app')
```

### Inception
```
class org.openrewrite.java.tree.JavaType$Parameterized cannot be cast to class org.openrewrite.java.tree.JavaType$Class (org.openrewrite.java.tree.JavaType$Parameterized and org.openrewrite.java.tree.JavaType$Class are in unnamed module of loader 'app')
```

### Generic
```
java.lang.ClassCastException: class org.openrewrite.java.tree.JavaType$GenericTypeVariable cannot be cast to class org.openrewrite.java.tree.JavaType$Class (org.openrewrite.java.tree.JavaType$GenericTypeVariable and org.openrewrite.java.tree.JavaType$Class are in unnamed module of loader 'app')
org.openrewrite.java.migrate.lang.var.UseVarForGenericsConstructors$UseVarForGenericsVisitor.lambda$transformToVar$0(UseVarForGenericsConstructors.java:121)
java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
...
```

### Unbound
```
java.lang.ClassCastException: class org.openrewrite.java.tree.JavaType$GenericTypeVariable cannot be cast to class org.openrewrite.java.tree.JavaType$Class (org.openrewrite.java.tree.JavaType$GenericTypeVariable and org.openrewrite.java.tree.JavaType$Class are in unnamed module of loader 'app')
org.openrewrite.java.migrate.lang.var.UseVarForGenericsConstructors$UseVarForGenericsVisitor.lambda$transformToVar$0(UseVarForGenericsConstructors.java:121)
java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
...
```

### Inception
```
java.lang.ClassCastException: class org.openrewrite.java.tree.JavaType$Parameterized cannot be cast to class org.openrewrite.java.tree.JavaType$Class (org.openrewrite.java.tree.JavaType$Parameterized and org.openrewrite.java.tree.JavaType$Class are in unnamed module of loader 'app')
org.openrewrite.java.migrate.lang.var.UseVarForGenericsConstructors$UseVarForGenericsVisitor.lambda$transformToVar$0(UseVarForGenericsConstructors.java:121)
java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
...
```

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?

yup.

**Comment by MBoegers** [comment]
It is fixed in #259, bounded generics e.q. `List<? extends String>` are skipped because I am unable to parse Bounds as Expressions.

**Comment by joanvr** [comment]
Thank you very much for your contribution @MBoegers! Let's follow the conversation in the PR
